apps/fz_wall/lib/fz_wall/cli/live.ex

```diff
@@ -6,104 +6,139 @@ defmodule FzWall.CLI.Live do
+  """
+  def add_device(device) do
+    get_types()
+    |> Enum.each(fn type -> add_to_set(device.user_id, device[type], type) end)
   end
 
   @doc """
-  List currently loaded rules.
+  Adds rule ip to its corresponding sets.
   """
-  def list_rules do
-    exec!("#{nft()} -a list table inet #{@table_name}")
+  def add_rule(rule) do
+    add_to_set(rule.user_id, rule.destination, proto(rule.destination), rule.action)
   end
 
   @doc """
-  Deletes nftables rule.
+  Delete rule destination ip from its corresponding sets.
   """
-  def delete_rule(params) do
-    rule_str(params)
-    |> delete_rule_matching()
+  def delete_rule(rule) do
+    remove_from_set(rule.user_id, rule.destination, proto(rule.destination), rule.action)
   end
 
-  def delete_rules(source) do
-    source_match_str(source)
-    |> delete_rule_matching()
+  @doc """
+  Eliminates device rules from its corresponding sets.
+  """
+  def delete_device(device) do
+    get_types()
+    |> Enum.each(fn type -> remove_from_set(device.user_id, device[type], type) end)
   end
 
-  defp delete_rule_matching(rule_str) do
-    rules = exec!("#{nft()} -a list table inet #{@table_name}")
+  defp remove_from_set(_user_id, nil, _type), do: :no_ip
 
-    # When a rule is deleted the others might change handle so we need to
-    # re-scan each time.
-    case rule_handle_regex(~r/^\s*#{rule_str}.*# handle (?<num>\d+)/m, rules) do
-      nil ->
-        :no_rule
+  defp remove_from_set(user_id, ip, type) do
+    get_device_set_name(user_id, type)
+    |> delete_elem(ip)
+  end
 
-      [handle] ->
-        exec!("#{nft()} delete rule inet #{@table_name} forward handle #{handle}")
-        # There might still be matching rules so re-run it
-        delete_rule_matching(rule_str)
-    end
+  defp remove_from_set(user_id, ip, type, action) do
+    get_dest_set_name(user_id, type, action)
+    |> delete_elem(ip)
   end
 
-  @doc """
-  Restores rules.
-  """
-  def restore(rules) do
-    # XXX: Priority?
-    for rule_spec <- rules do
-      add_rule(rule_spec)
-    end
+  defp add_to_set(_user_id, nil, _type), do: :no_ip
+
+  defp add_to_set(user_id, ip, type) do
+    get_device_set_name(user_id, type)
+    |> add_elem(ip)
+  end
+
+  defp add_to_set(user_id, ip, type, action) do
+    get_dest_set_name(user_id, type, action)
+    |> add_elem(ip)
+  end
+
+  defp add_sets(user_id) do
+    list_sets(user_id)
+    |> Enum.each(fn set -> add_set(set) end)
+  end
+
+  defp delete_sets(user_id) do
+    list_sets(user_id)
+    |> Enum.each(fn set -> delete_set(set) end)
+  end
+
+  defp add_rules(user_id) do
+    cross(get_types(), get_actions())
+    |> Enum.each(fn {type, action} ->
+      create_rule(
+        type,
+        get_device_set_name(user_id, type),
+        get_dest_set_name(user_id, type, action),
+        action
+      )
+    end)
+  end
+
+  defp delete_rules(user_id) do
+    cross(get_types(), get_actions())
+    |> Enum.each(fn {type, action} ->
+      remove_rule(
```

Comment:
Would it be worth calling the `list table` exec outside this loop? Probably can get away with it for now, but doing the `regex` across the whole table for each rule to delete might not be ideal?

Or does the handle change after each deletion?

Reply:
The handle **can** change after each deletion. When I tested with it outside the loop it worked but in some manual tests I did the handles changed and I couldn't find anything in the docs that guarantees the same handle after deletion/additions so I choose to keep it this way.